rfcs/0005-pablo-distribution-assets/0005-pablo-distribution.adoc

```diff
@@ -0,0 +1,632 @@
+
+=== PICA Governance
+. Governance MUST be able to set the PICA token reward allocation.
+. Governance MUST be able to adjust the PICA reward rate based on the incentivization strategy.
+
+=== Technical Requirements
+. The system MUST allow accumulation and mapping of rewards shares of multiple assets types(Eg: PBLO, KSM) to staked position(fNFT) type defined by another asset type(eg: PICA).
+. The system MUST support transfer of rewards using staking-rewards pallet to necessary fNFT types.
+. The system SHOULD support converting a reward accumulated in one asset type to another based on a preferred reward asset type configuration. Eg: Given a reward accumulated is in Acala it should be able to convert that to one of PBLO or PICA using the Pablo DEX pools.
+** This is to handle cases where a Pablo pool fees are in a different asset type than what is preferred.
+
+== Method
+
+=== System Overview
+
+[plantuml,images/pablo-distribution-verview,png]
+----
+skinparam component {
+  backgroundColor<<exists>> LightGreen
+  backgroundColor<<remove>> Red
+}
+
+node "Pallet-Staking-Rewards\n(Pallet-Rewards?)" {
+    StakingReward - [Reward Logic]
+    Staking - [fNFT Storage/Logic]
+    Node "BatchProcess" {
+        [RewardDisbursementHook]
+        [RewardsConfig] --> [RewardAccumulationHook]
+    }
+}
+
+[fNFT Storage/Logic] <<exists>>
+[RewardDisbursementHook] <<remove>>
+
+node "Pallet-Pablo" {
+    [Pool] --> StakingReward : transfer\ntrading fees\nfor PBLO stakers
+    [Pool] --> Staking : stake LP tokens
+    [Pool] --> [Pool] : accumulate LP\ntrading fees
+    [FeeConfig] --> [Pool]
+}
+
+[Pool] <<exists>>
+
+
+node "Governance" {
+    [Configuration] --> [FeeConfig]
+    [Treasury]--> StakingReward: Transfer LP, PBLO\nor other token rewards
+}
+----
+
+TODO: What to do for part of protocol fees that should be transferred to treasury eventually as treasury does not stake it's PBLO?
+
+=== Pallet-Pablo
+
+In order to 1. support LP staking 2. LP trading fee distribution and 3. PBLO staking reward using trading fees, following changes are proposed for https://github.com/ComposableFi/composable/tree/main/frame/pablo[Pallet-Pablo].
+
+[#_feeconfig]
+==== FeeConfig
+
+Each pool in Pablo defines a fee percentage to be charged for each trade.Except for LBPs other pools also define an owner fee that is a percentage out of the main trading fee. The `FeeConfig` is a new abstraction over all fees that could be charged on a pool to allow for extension. At this time a 100% of the owner fee should be defined as a new field `protocol_fee`.
+
+[plantuml,images/pablo-fee-config,png]
+----
+class FeeConfig {
+    +fee_rate: Permill
+    +owner_fee_rate: Permill,
+    +protocol_fee_rate: Permill,
+}
+----
+
+*Existing code must be modified to use this data structure*.
+
+Given this,
+----
+fee = // calculation depends on the pool type: based on the fee_rate
```

Comment:
please make this python

```

owner_fee_rate = 20/100
protocol_fee_rate = 100/100

fee = 100
owner_fee = fee * owner_fee_rate * (1 - protocol_fee_rate)
protocol_fee = owner_fee * protocol_fee_rate

print(protocol_fee)
print(owner_fee)
```
I reject PR until it is fixed on why both fees are 0

Reply:
```
protocol_fee = fee * owner_fee_rate * (1 - protocol_fee_rate) * protocol_fee_rate


y = C (1 - x) * x

y = C x - C x ^ 2

optimal,

C = 2 C x  

x = 1/ 2
```

maximal protocol amount will be in case of its fee is 0.5

Reply:
so owner_fee goes to PBLO holder, if its is zero then nothing goes to them.
AFAIU protocol_fee is value transferred to liquidity providers. So having protocol_fee_rate zero makes sense only perhaps for LBP. 
So Fee= protocol_fee - (protocol_fee * owner_fee_percent)

Reply:
`owner_fee_rate` is intended to capture the total fee that goes out to owner plus the protocol token (PBLO) holders,

as default for this can be found already implemented at https://github.com/ComposableFi/composable/blob/main/frame/composable-traits/src/dex.rs#L175

Reply:
The idea is that PBLO owners would only get a cut of the fee of the owner of the pool, if they don't charge, we don't charge. But by default the pool owner fee is zero and PBLO holders get the full owner fee.